### PR TITLE
AP_RangeFinder: [AP_RangeFinder_BLPing] Update includes and limit max update rate

### PR DESCRIPTION
**Pull request by patrickelectric**
@jaxxzer @rmackay9 @peterbarker

### CODE REVIEW

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp [hunk 1]
```diff
@@ -78,6 +81,11 @@ void AP_RangeFinder_BLPing::update(void)
 
     const uint32_t now = AP_HAL::millis();
 
+    // Limit time between update requests
+    if (now - state.last_reading_ms < BLPING_MIN_UPDATE_TIME_MS) {
```
Comment (OXINARF) [inline libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: I'm quite confused by this. If there are new bytes to read why not read them?

It looks like the issue might be sending the request for the sensor (which, in any case, should handled in the sending part, not the reading one), but it's also not clear to me if that is actually an issue - does something bad happens when requesting it too fast? I would imagine that if the sensor didn't update yet then it will return the last value, but I don't see what would be wrong with that (there is a possibility the sensor doesn't behave well with this, I don't know, that's basically my question 🙂).

Comment (patrickelectric) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: The problem here is that `get_reading` is the one that do the message request, so that's why this is being done here.

About the sensor, the fastest frequency that the sensor can operate is around 30Hz. When operating in the max distance (70m) the time between each profile will be around 10ms. Sending requests at 50Hz will result in not periodically results and in a poor performance.
Imagine that you request 50 messages in 1 second, since the sensor has a buffer for 3~4 messagesm you'll receive something like:
3x Profiles...
20x Nothing...
3x Profiles...
20x Nothing...
3x Profiles...
In other words, the cpu will be sending unnecessary messages and getting only 3 distances in 1 second and not 10 different distances.

Comment (jaxxzer) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: It sounds like the problem is that the device firmware cannot tolerate the data requests at the current 50Hz (it should, but alas).

Comment (patrickelectric) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: Yes, this should be fixed in newer versions of the firmware.

Comment (OXINARF) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: > When operating in the max distance (70m) the time between each profile will be around 10ms. 

@patrickelectric You meant 100ms right?

> About the sensor, the fastest frequency that the sensor can operate is around 30Hz.

So why can't we send messages at 30Hz? I think I might be confused by what your comment "When operating in the max distance" actually means - I assumed that was worst case scenario, but maybe that's a fixed setting? In that case, the sensor would operate at 10Hz, not 30 (hence my confusion).

Sorry, not trying to delay this, just trying to understand it. In any case, I'm still of the opinion that this should block sending the request message and not the entire reading.

Comment (jaxxzer) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: > So why can't we send messages at 30Hz?

@OXINARF not sure if this is what you are asking, but there is a firmware bug imposing a limitation on the maximum request rate.

> In any case, I'm still of the opinion that this should block sending the request message and not the entire reading.

@patrickelectric I agree with this.

**Comment by jaxxzer** [comment]
@patrickelectric the requests are already being rate-limited:
[here](https://github.com/ArduPilot/ardupilot/pull/11321/files#diff-808bc25d8005172cc2ec3a3b7de2fac4R181) and [here](https://github.com/ArduPilot/ardupilot/pull/11321/files#diff-808bc25d8005172cc2ec3a3b7de2fac4R97)

This might need some refactoring rather than another bit of rate-limiting logic.

If the bug/fix actually applies to the device firmware, then we can close this.

We should get a logic trace to understand exactly what is happening.

File: libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp (hunk 1 above)
Comment (OXINARF) [reply libraries/AP_RangeFinder/AP_RangeFinder_BLPing.cpp]: @jaxxzer What I'm basically asking is what is the rate it is limited to. From @patrickelectric's answer I thought it was 30Hz, but if that is it, then why limit the driver to send the message at 10Hz?
